protocols/manage/propagate.go

```diff
@@ -51,22 +47,48 @@ type PropagateReply struct {
 	Level int
 }
 
-// PropagateStartAndWait starts the propagation protocol and blocks until
+// PropagationFunc starts the propagation protocol and blocks until
 // all children stored the new value or the timeout has been reached.
 // The return value is the number of nodes that acknowledged having
 // stored the new value or an error if the protocol couldn't start.
-func PropagateStartAndWait(c *sda.Context, el *sda.Roster, msg network.Body, msec int, f func(network.Body)) (int, error) {
-	tree := el.GenerateNaryTreeWithRoot(8, c.ServerIdentity())
-	log.Lvl3("Starting to propagate", reflect.TypeOf(msg))
-	pi, err := c.CreateProtocolService("Propagate", tree)
-	if err != nil {
-		return -1, err
-	}
-	return propagateStartAndWait(pi, msg, msec, f)
+type PropagationFunc func(el *sda.Roster, msg network.Body, msec int) (int, error)
+
+// PropagationStore is the function that will store the new data.
+type PropagationStore func(network.Body)
+
+// NewPropagationFunc registers a new protocol name with the context c and will
+// set f as handler for every new instance of that protocol.
+func NewPropagationFunc(c *sda.Context, name string, f PropagationStore) (PropagationFunc, error) {
+	pid, err := c.ProtocolRegister(name, func(n *sda.TreeNodeInstance) (sda.ProtocolInstance, error) {
+		p := &Propagate{
+			sd:               &PropagateSendData{[]byte{}, 1000},
+			TreeNodeInstance: n,
+			received:         0,
+			subtreeCount:     n.TreeNode().SubtreeCount(),
+			onData:           f,
+		}
+		for _, h := range []interface{}{&p.ChannelSD, &p.ChannelReply} {
+			if err := p.RegisterChannel(h); err != nil {
+				return nil, err
+			}
+		}
+		return p, nil
+	})
+	log.Lvl3("Registering new propagation for", c.ServerIdentity(),
+		name, pid)
+	return func(el *sda.Roster, msg network.Body, msec int) (int, error) {
+		tree := el.GenerateNaryTreeWithRoot(8, c.ServerIdentity())
+		log.Lvl3(el.List[0].Address, "Starting to propagate", reflect.TypeOf(msg))
+		pi, err := c.CreateProtocolSDA(name, tree)
+		if err != nil {
+			return -1, err
+		}
+		return propagateStartAndWait(pi, msg, msec, f)
+	}, err
```

Comment:
Maybe I'm missing something, but it is difficult for me to see how all these changes in the propagate protocol are related/necessary because of #656. Can you explain it here or preferably in the issue? 
Currently, it just says: 
> Now that each conode has its own set of Protocols, the ProtocolRegister method should also be made available to services.